{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled1.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "qFvTew1BdRdT"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ssBoSBbggEnJ"
      },
      "source": [
        "https://www.ncbi.nlm.nih.gov/genome/154?genome_assembly_id=299272\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PlbrF9tejc0W",
        "outputId": "fb165cf6-9251-4a52-c9ec-17f0dc1207a6"
      },
      "source": [
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Staphylococcus_aureus/reference/GCF_000013425.1_ASM1342v1/GCF_000013425.1_ASM1342v1_genomic.fna.gz\n",
        "!gzip -d GCF_000013425.1_ASM1342v1_genomic.fna.gz "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-09-06 15:36:37--  https://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Staphylococcus_aureus/reference/GCF_000013425.1_ASM1342v1/GCF_000013425.1_ASM1342v1_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.12, 2607:f220:41e:250::12, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 820675 (801K) [application/x-gzip]\n",
            "Saving to: ‘GCF_000013425.1_ASM1342v1_genomic.fna.gz’\n",
            "\n",
            "GCF_000013425.1_ASM 100%[===================>] 801.44K  2.11MB/s    in 0.4s    \n",
            "\n",
            "2021-09-06 15:36:38 (2.11 MB/s) - ‘GCF_000013425.1_ASM1342v1_genomic.fna.gz’ saved [820675/820675]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RDHA-Oi_gIS_",
        "outputId": "3812e3b3-8d99-430b-f129-75029fd4663f"
      },
      "source": [
        "!chmod a+x zhunt2\n",
        "!./zhunt2 12 8 12 GCF_000013425.1_ASM1342v1_genomic.fna"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "dinucleotides 12\n",
            "min/max 8 12\n",
            "min/max 8 12\n",
            "operating on GCF_000013425.1_ASM1342v1_genomic.fna\n",
            "calculating zscore\n",
            "opening GCF_000013425.1_ASM1342v1_genomic.fna\n",
            "inputting sequence\n",
            "opening GCF_000013425.1_ASM1342v1_genomic.fna.Z-SCORE\n",
            "\n",
            " run time=634 sec\n",
            "use min/max 8 12\n",
            "analyzing_zscore\n",
            "opening GCF_000013425.1_ASM1342v1_genomic.fna.Z-SCORE\n",
            "opening GCF_000013425.1_ASM1342v1_genomic.fna\n",
            "inputting sequence\n",
            "GCF_000013425.1_ASM1342v1_genomic.fna 2821379 8 12\n",
            "1 17 16  30.238  33.580 4.149510e-01 nctacccaanctccct   ASASASASSASASASA\n",
            "2 18 16  30.919  24.115 3.208749e-01 ctacccaanctccctg   SASASASAASASASAS\n",
            "3 21 18  30.150  33.147 4.275896e-01 tacccaanctccctgncg   ASASASSASASASAASAS\n",
            "4 20 16  31.166  23.305 2.891363e-01 acccaanctccctgnc   SASASAASASASASAS\n",
            "5 21 16  30.568  33.169 3.684349e-01 cccaanctccctgncg   ASASSASASASAASAS\n",
            "6 22 16  31.799  17.175 2.151047e-01 ccaanctccctgncga   ASASASASASASASAS\n",
            "7 25 18  31.600  30.332 2.371638e-01 caanctccctgncgatta   ASSASASASAASASASAS\n",
            "8 24 16  31.840  16.241 2.108015e-01 aanctccctgncgatt   ASASASASASASASAS\n",
            "9 25 16  32.072  21.608 1.870246e-01 anctccctgncgatta   SASASASAASASASAS\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hciqUx9rqAvn",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "97bc1bc7-6eb4-448c-c321-830af61084ec"
      },
      "source": [
        "import pandas as pd\n",
        "\n",
        "z = pd.read_csv(\n",
        "    \"GCF_000013425.1_ASM1342v1_genomic.fna.Z-SCORE\",\n",
        "    skiprows=1,\n",
        "    names=[\"start\", \"end\", \"1\", \"2\", \"3\", \"score\", \"seq\", \"4\"],\n",
        "    delim_whitespace=True\n",
        "    )\n",
        "\n",
        "ZSCORE_MIN = 300\n",
        "res = z.loc[z.score > ZSCORE_MIN]\n",
        "zfile = open(\"zhunt_res.bed\", \"w\")\n",
        "print(res.columns)\n",
        "for idx,row in res.iterrows():\n",
        "    zfile.write(f'NC_007795.1\\t{row[\"start\"]}\\t{row[\"end\"]}\\n')\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index(['start', 'end', '1', '2', '3', 'score', 'seq', '4'], dtype='object')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Jx1muqD8qHGu",
        "outputId": "fde6afc0-9d21-437b-c9b0-5bf2d3822a76"
      },
      "source": [
        "!wget https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools.static.binary\n",
        "!mv bedtools.static.binary bedtools\n",
        "!chmod a+x bedtools\n",
        "\n",
        "\n",
        "!wget https://github.com/bedops/bedops/releases/download/v2.4.39/bedops_linux_x86_64-v2.4.39.tar.bz2\n",
        "!tar jxvf bedops_linux_x86_64-v2.4.39.tar.bz2\n",
        "!cp bin/* /usr/local/bin"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-09-06 15:58:37--  https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools.static.binary\n",
            "Resolving github.com (github.com)... 192.30.255.112\n",
            "Connecting to github.com (github.com)|192.30.255.112|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-releases.githubusercontent.com/15059334/c633cf80-61f8-11eb-92ef-18b90dff37e2?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210906%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210906T155838Z&X-Amz-Expires=300&X-Amz-Signature=9f46da5c151d5254b6906d11e21f51a11b2e9efd2ac85bef703bf74dbee875a1&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools.static.binary&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-09-06 15:58:38--  https://github-releases.githubusercontent.com/15059334/c633cf80-61f8-11eb-92ef-18b90dff37e2?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210906%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210906T155838Z&X-Amz-Expires=300&X-Amz-Signature=9f46da5c151d5254b6906d11e21f51a11b2e9efd2ac85bef703bf74dbee875a1&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools.static.binary&response-content-type=application%2Foctet-stream\n",
            "Resolving github-releases.githubusercontent.com (github-releases.githubusercontent.com)... 185.199.108.154, 185.199.109.154, 185.199.110.154, ...\n",
            "Connecting to github-releases.githubusercontent.com (github-releases.githubusercontent.com)|185.199.108.154|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 41993576 (40M) [application/octet-stream]\n",
            "Saving to: ‘bedtools.static.binary’\n",
            "\n",
            "bedtools.static.bin 100%[===================>]  40.05M  17.9MB/s    in 2.2s    \n",
            "\n",
            "2021-09-06 15:58:40 (17.9 MB/s) - ‘bedtools.static.binary’ saved [41993576/41993576]\n",
            "\n",
            "--2021-09-06 15:58:40--  https://github.com/bedops/bedops/releases/download/v2.4.39/bedops_linux_x86_64-v2.4.39.tar.bz2\n",
            "Resolving github.com (github.com)... 192.30.255.113\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-releases.githubusercontent.com/12932856/46ea9f00-7866-11ea-8a7c-c01e99a1339e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210906%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210906T155840Z&X-Amz-Expires=300&X-Amz-Signature=a3293dc5ca87b79d7f9d902e507c8732f26039293f53342059e7d31c4ae87cf0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=12932856&response-content-disposition=attachment%3B%20filename%3Dbedops_linux_x86_64-v2.4.39.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-09-06 15:58:41--  https://github-releases.githubusercontent.com/12932856/46ea9f00-7866-11ea-8a7c-c01e99a1339e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210906%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210906T155840Z&X-Amz-Expires=300&X-Amz-Signature=a3293dc5ca87b79d7f9d902e507c8732f26039293f53342059e7d31c4ae87cf0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=12932856&response-content-disposition=attachment%3B%20filename%3Dbedops_linux_x86_64-v2.4.39.tar.bz2&response-content-type=application%2Foctet-stream\n",
            "Resolving github-releases.githubusercontent.com (github-releases.githubusercontent.com)... 185.199.108.154, 185.199.109.154, 185.199.110.154, ...\n",
            "Connecting to github-releases.githubusercontent.com (github-releases.githubusercontent.com)|185.199.108.154|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 25646650 (24M) [application/octet-stream]\n",
            "Saving to: ‘bedops_linux_x86_64-v2.4.39.tar.bz2’\n",
            "\n",
            "bedops_linux_x86_64 100%[===================>]  24.46M  81.9MB/s    in 0.3s    \n",
            "\n",
            "2021-09-06 15:58:41 (81.9 MB/s) - ‘bedops_linux_x86_64-v2.4.39.tar.bz2’ saved [25646650/25646650]\n",
            "\n",
            "bin/\n",
            "bin/bam2bed_sge\n",
            "bin/update-sort-bed-migrate-candidates-float128\n",
            "bin/update-sort-bed-migrate-candidates-typical\n",
            "bin/bam2starch_sge-float128\n",
            "bin/psl2starch-typical\n",
            "bin/bam2starch_gnuParallel\n",
            "bin/bam2starch_slurm\n",
            "bin/gtf2starch-typical\n",
            "bin/gvf2bed\n",
            "bin/wig2bed-float128\n",
            "bin/starchcluster_sge\n",
            "bin/unstarch\n",
            "bin/update-sort-bed-starch-slurm-megarow\n",
            "bin/update-sort-bed-migrate-candidates\n",
            "bin/psl2starch\n",
            "bin/bedextract-float128\n",
            "bin/vcf2bed\n",
            "bin/psl2bed-megarow\n",
            "bin/bam2bed_gnuParallel-megarow\n",
            "bin/gvf2bed-typical\n",
            "bin/gtf2bed-typical\n",
            "bin/gff2bed-typical\n",
            "bin/switch-BEDOPS-binary-type\n",
            "bin/sam2starch-float128\n",
            "bin/starchcluster_slurm-megarow\n",
            "bin/gvf2starch-typical\n",
            "bin/bedops-float128\n",
            "bin/unstarch-megarow\n",
            "bin/update-sort-bed-slurm-float128\n",
            "bin/bedmap-megarow\n",
            "bin/bam2starch_slurm-float128\n",
            "bin/starch-float128\n",
            "bin/starchcluster_gnuParallel\n",
            "bin/sort-bed-megarow\n",
            "bin/convert2bed-megarow\n",
            "bin/update-sort-bed-starch-slurm\n",
            "bin/update-sort-bed-starch-slurm-float128\n",
            "bin/sam2bed-float128\n",
            "bin/bam2bed_gnuParallel\n",
            "bin/bam2starch_slurm-typical\n",
            "bin/bam2starch\n",
            "bin/bam2bed_slurm-megarow\n",
            "bin/bedextract-megarow\n",
            "bin/gvf2starch-float128\n",
            "bin/closest-features-float128\n",
            "bin/bam2starch_gnuParallel-typical\n",
            "bin/bedops-megarow\n",
            "bin/bam2starch_sge-typical\n",
            "bin/rmsk2bed-float128\n",
            "bin/gtf2bed-megarow\n",
            "bin/gff2bed-float128\n",
            "bin/unstarch-float128\n",
            "bin/rmsk2starch\n",
            "bin/convert2bed-float128\n",
            "bin/psl2starch-megarow\n",
            "bin/psl2starch-float128\n",
            "bin/bedmap-typical\n",
            "bin/bam2bed_slurm-float128\n",
            "bin/starchstrip-megarow\n",
            "bin/update-sort-bed-slurm-typical\n",
            "bin/convert2bed\n",
            "bin/starchstrip-float128\n",
            "bin/bam2starch_gnuParallel-megarow\n",
            "bin/gff2starch-float128\n",
            "bin/wig2starch-typical\n",
            "bin/starchcat\n",
            "bin/gff2starch\n",
            "bin/bam2starch-megarow\n",
            "bin/sam2starch-megarow\n",
            "bin/starchcluster_slurm-float128\n",
            "bin/sort-bed-typical\n",
            "bin/update-sort-bed-slurm-megarow\n",
            "bin/gvf2bed-megarow\n",
            "bin/starchcluster_slurm\n",
            "bin/bam2bed_gnuParallel-float128\n",
            "bin/starch-megarow\n",
            "bin/bam2starch_slurm-megarow\n",
            "bin/starchcat-typical\n",
            "bin/bedmap\n",
            "bin/sam2bed-megarow\n",
            "bin/gtf2bed\n",
            "bin/bam2starch_gnuParallel-float128\n",
            "bin/gff2bed-megarow\n",
            "bin/starch-diff-megarow\n",
            "bin/sam2bed\n",
            "bin/bam2bed-megarow\n",
            "bin/bam2bed-float128\n",
            "bin/starch-typical\n",
            "bin/bam2bed_slurm-typical\n",
            "bin/vcf2bed-megarow\n",
            "bin/closest-features-megarow\n",
            "bin/bedmap-float128\n",
            "bin/starchcluster_gnuParallel-typical\n",
            "bin/rmsk2starch-typical\n",
            "bin/bam2bed-typical\n",
            "bin/convert2bed-typical\n",
            "bin/sort-bed-float128\n",
            "bin/vcf2starch-megarow\n",
            "bin/closest-features-typical\n",
            "bin/gvf2starch\n",
            "bin/wig2bed\n",
            "bin/sam2starch-typical\n",
            "bin/bam2bed_slurm\n",
            "bin/bam2bed_sge-megarow\n",
            "bin/bam2starch_sge\n",
            "bin/gvf2starch-megarow\n",
            "bin/rmsk2bed\n",
            "bin/wig2starch\n",
            "bin/starchcluster_gnuParallel-float128\n",
            "bin/starch-diff-typical\n",
            "bin/starch\n",
            "bin/gtf2starch\n",
            "bin/starchcluster_slurm-typical\n",
            "bin/psl2bed-typical\n",
            "bin/sort-bed\n",
            "bin/rmsk2bed-typical\n",
            "bin/bam2bed_sge-float128\n",
            "bin/bam2starch-typical\n",
            "bin/starch-diff\n",
            "bin/starchcluster_sge-float128\n",
            "bin/starchcluster_sge-megarow\n",
            "bin/starchcat-megarow\n",
            "bin/wig2bed-megarow\n",
            "bin/gff2starch-megarow\n",
            "bin/vcf2starch-float128\n",
            "bin/bam2bed\n",
            "bin/bedextract-typical\n",
            "bin/gff2starch-typical\n",
            "bin/starchcat-float128\n",
            "bin/bam2bed_sge-typical\n",
            "bin/wig2starch-float128\n",
            "bin/starchstrip\n",
            "bin/bam2starch-float128\n",
            "bin/gtf2bed-float128\n",
            "bin/bedops-typical\n",
            "bin/psl2bed\n",
            "bin/gff2bed\n",
            "bin/starchcluster_gnuParallel-megarow\n",
            "bin/starch-diff-float128\n",
            "bin/vcf2bed-typical\n",
            "bin/gtf2starch-float128\n",
            "bin/bedops\n",
            "bin/sam2starch\n",
            "bin/bam2starch_sge-megarow\n",
            "bin/wig2bed-typical\n",
            "bin/gtf2starch-megarow\n",
            "bin/vcf2starch-typical\n",
            "bin/psl2bed-float128\n",
            "bin/rmsk2starch-megarow\n",
            "bin/rmsk2bed-megarow\n",
            "bin/unstarch-typical\n",
            "bin/vcf2bed-float128\n",
            "bin/update-sort-bed-slurm\n",
            "bin/rmsk2starch-float128\n",
            "bin/bam2bed_gnuParallel-typical\n",
            "bin/update-sort-bed-starch-slurm-typical\n",
            "bin/update-sort-bed-migrate-candidates-megarow\n",
            "bin/starchcluster_sge-typical\n",
            "bin/sam2bed-typical\n",
            "bin/vcf2starch\n",
            "bin/gvf2bed-float128\n",
            "bin/closest-features\n",
            "bin/wig2starch-megarow\n",
            "bin/starchstrip-typical\n",
            "bin/bedextract\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bZn2wJr4XjBG",
        "outputId": "7053fa21-3c16-4db8-c9a1-577979fed731"
      },
      "source": [
        "!rm genomic.bed\n",
        "!rm GCF_000013425.1_ASM1342v1_genomic.gff\n",
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Staphylococcus_aureus/reference/GCF_000013425.1_ASM1342v1/GCF_000013425.1_ASM1342v1_genomic.gff.gz\n",
        "!gzip -d GCF_000013425.1_ASM1342v1_genomic.gff.gz\n",
        "!gff2bed < GCF_000013425.1_ASM1342v1_genomic.gff > genomic.bed\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "rm: cannot remove 'genomic.bed': No such file or directory\n",
            "rm: cannot remove 'GCF_000013425.1_ASM1342v1_genomic.gff': No such file or directory\n",
            "--2021-09-06 15:58:55--  https://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Staphylococcus_aureus/reference/GCF_000013425.1_ASM1342v1/GCF_000013425.1_ASM1342v1_genomic.gff.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 165.112.9.229, 130.14.250.13, 2607:f220:41e:250::12, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|165.112.9.229|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 197738 (193K) [application/x-gzip]\n",
            "Saving to: ‘GCF_000013425.1_ASM1342v1_genomic.gff.gz’\n",
            "\n",
            "GCF_000013425.1_ASM 100%[===================>] 193.10K  1.06MB/s    in 0.2s    \n",
            "\n",
            "2021-09-06 15:58:56 (1.06 MB/s) - ‘GCF_000013425.1_ASM1342v1_genomic.gff.gz’ saved [197738/197738]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IHGReDHF-sVK"
      },
      "source": [
        "!./bedtools intersect -a genomic.bed -b zhunt_res.bed -wa > overlaps.bed"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uA1wRowwAFVE",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "63de9e86-10b0-4659-ee83-b55c72914b74"
      },
      "source": [
        " !uniq overlaps.bed | grep -o .*product=.* > zres.csv\n",
        " z = pd.read_csv(\n",
        "    \"zres.csv\",\n",
        "    skiprows=0,\n",
        "    names=[\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\"],\n",
        "    sep=\"\\t\",\n",
        "    )\n",
        " print(z[\"8\"].value_counts())"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CDS     320\n",
            "exon      8\n",
            "rRNA      5\n",
            "tRNA      3\n",
            "Name: 8, dtype: int64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UjMxe4zvA3Jx",
        "outputId": "2e9ff11f-9c5c-4620-e247-4be8a585a716"
      },
      "source": [
        "!pip install Bio\n",
        "from Bio import SeqIO\n",
        "import re"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: Bio in /usr/local/lib/python3.7/dist-packages (0.8.5)\n",
            "Requirement already satisfied: mygene in /usr/local/lib/python3.7/dist-packages (from Bio) (3.2.2)\n",
            "Requirement already satisfied: biopython>=1.79 in /usr/local/lib/python3.7/dist-packages (from Bio) (1.79)\n",
            "Requirement already satisfied: tqdm>=4.62 in /usr/local/lib/python3.7/dist-packages (from Bio) (4.62.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.23.0)\n",
            "Requirement already satisfied: plac in /usr/local/lib/python3.7/dist-packages (from Bio) (1.1.3)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.19.5)\n",
            "Requirement already satisfied: biothings-client>=0.2.6 in /usr/local/lib/python3.7/dist-packages (from mygene->Bio) (0.2.6)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2021.5.30)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (3.0.4)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "metadata": {},
          "execution_count": 27
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "E111uNnn4_QX",
        "outputId": "c458ee89-3640-42bc-be02-9d508f19f89f"
      },
      "source": [
        "pattern=\"(?:G{2,}[ATGC]{1,7}){2,}G{2,}\"\n",
        "\n",
        "for record in SeqIO.parse(\"GCF_000013425.1_ASM1342v1_genomic.fna\", \"fasta\"):\n",
        "  print(\"%s %i\" % (record.id, len(record)))\n",
        "  name, sequence = record.id, str(record.seq)\n",
        "PQS=[[m.group(0), m.start(),m.end()] for m in re.finditer(pattern,sequence)]\n",
        "print(len(PQS))\n",
        "zfile = open(\"quadruplex_res.bed\", \"w\")\n",
        "print(res.columns)\n",
        "for idx,row in res.iterrows():\n",
        "    zfile.write(f'NC_007795.1\\t{row[\"start\"]}\\t{row[\"end\"]}\\n')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NC_007795.1 2821361\n",
            "1957\n",
            "Index(['start', 'end', '1', '2', '3', 'score', 'seq', '4'], dtype='object')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "pvNGopFW61iQ"
      },
      "source": [
        "!./bedtools intersect -a genomic.bed -b quadruplex_res.bed -wa > overlaps2.bed"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0oOcpkQU7Zuc",
        "outputId": "c7878d33-d1dc-487f-a8a9-b840ef735e68"
      },
      "source": [
        " \n",
        " !uniq overlaps2.bed  > qres.csv\n",
        " q = pd.read_csv(\n",
        "    \"qres.csv\",\n",
        "    skiprows=1,\n",
        "    names=[\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\"],\n",
        "    sep=\"\\t\",\n",
        "    )\n",
        " print(len(q))\n",
        " print(z[\"8\"].value_counts())"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "683\n",
            "gene          336\n",
            "CDS           328\n",
            "exon            8\n",
            "rRNA            5\n",
            "tRNA            3\n",
            "pseudogene      3\n",
            "Name: 8, dtype: int64\n"
          ]
        }
      ]
    }
  ]
}